File: 程序员面试金典/8.6汉诺塔问题/hanota.py
class Solution:
    def hanota(self, A: List[int], B: List[int], C: List[int]) -> None:
        """
        Do not return anything, modify C in-place instead.
        """
        self.hanoi(A, B, C, len(A))
    
    def hanoi(self, A, B, C, n):
        if n == 1:
            C.append(A.pop())
        else:
            self.hanoi(A, C, B, n - 1)
            C.append(A.pop())
            self.hanoi(B, A, C, n - 1)

File: 程序员面试金典/8.6/hanota.java
class Solution {
    public void hanota(List<Integer> A, List<Integer> B, List<Integer> C) {
        hanoi(A, B, C, A.size());
    }

    private void hanoi(List<Integer> A, List<Integer> B, List<Integer> C, int n) {
        if (n == 1)
            C.add(A.remove(A.size() - 1));
        else {
            hanoi(A, C, B, n - 1);
            C.add(A.remove(A.size() - 1));
            hanoi(B, A, C, n - 1);
        }
    }

}
